<a href="https://colab.research.google.com/github/JMCastle98/M4_Analisis_de_Datos_con_Python/blob/main/Postwork01/Postwork01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Fe de erratas 🙏: Cambio de tema del proyecto final

<p align="center">
<img src="https://media1.tenor.com/images/f4dfba6861f151917aa8d895f19b84d9/tenor.gif?itemid=16694846"  height ="200" width = "300">
<p>

Originalmente nos plantemos realizar un proyecto para analizar datos sobre insuficiencia cardíaca y buscar un modelo de predicción utilizando el dataset [Hear Failure Prediction](https://www.kaggle.com/andrewmvd/heart-failure-clinical-data) del usuario Larxel en Kaggle. Al principio pudimos obtener ciertos resultados con estimados númericos y gráficos de distribución. Sin embargo, el dataset tiene ciertas limitantes como:

- Pocos datos: Menos de 300 registros
- Muchas variables categóricas binarias: De los 13 campos que incluye el dataset, 6 son de tipo categórico binario y esto dificultaba ó entorpecia algunas herramientas de análisis como las tablas de contingencia.
- Poca correlación entre las variables: A través del uso de un heatmap nos encontramos con que nuestra mejor correlación era de 0.16 y -0.19.

Considerando estos puntos, nos desaninamos un poco y decidimos buscar otros datasets, pero todos eran muy parecidos al que ya teniamos. Después de meditarlo un poco como equipo, decidimos cambiar el tema de nuestro proyecto, pasamos por diferentes opciones como [Análisis de criptomonedas](https://www.kaggle.com/sudalairajkumar/cryptocurrencypricehistory), [Datos de CPU's y GPU's](https://www.kaggle.com/iliassekkaf/computerparts) y [Riesgo de defunción en empleos de México](http://www.stps.gob.mx/gobmx/estadisticas/riesgos.htm). Todos eran temas super interesantes, pero no podiamos explotarlos con las herramientas que aprendimos en el módulo. 

Finalmente, a modo de curiosidad, revisamos un dataset de [pokémon](https://www.kaggle.com/mariotormo/complete-pokemon-dataset-updated-090420) que había sido actualizado muy recientemente. Parecía una idea algo descabellada, pero conforme revisamos los datos que contenia se nos empezaron a ocurrir varias ideas, de igual forma recordamos que en el módulo pasado trabajamos con la [PokéAPI](https://pokeapi.co/), nuestros proyectos anteriores habían estado relacionados con videojuegos y este año la franquicia celebra su 25 aniversario, así que... ¿Por qué no intentarlo? De esta manera nos propusimos realizar un análisis de los datos que se pueden encontrar en la Pokédex y convertirnos en los mejores **~~maestros pokémon~~** analistas de datos , ¡mejores que nadie más!.


<p align="center">
<img src="https://qph.fs.quoracdn.net/main-qimg-609283f9a315b4fc606c8c0e1c00071b" height ="200" width = "300">
<p>




# Hablemos sobre pokémon

La franquicia no necesita mucha presentación, desde febrero del año 1996, el juego de capturar criaturas y luchas junto a ellas para convertirte en el mejor se convirtió en toda una sensación. A lo largo de los años ha incurrido en diversas adaptaciones comos series animadas, mangas, cartas coleccionables y juguetes.

<p align="center">
<img src="https://images.ctfassets.net/ycci6h8ksgtu/274nX0UsGgY2R66HKxCZQT/3c6c622e4bb4c0db93149285e49a16f9/25th_promo_2x_optimized.png"  height ="200" width = "250">
<p>

La franquicia siempre se ha enfocado principalmente en los videojuegos, desde las primeras entregas para la Nintendo  Gameboy: Pokémon Red y Pokémon Green hasta las actuales versiones para Nintendo Switch: Pokémon Sword y Pokémon Shield han buscado mantener la esencia de la saga al incluir nuevas mecánicas de combate y sobretodo nuevos pokémon que llamaran la atención de los niños.

<p align="center">
<img src="https://cdn.i-scmp.com/sites/default/files/styles/og_image_young_post/public/d8/yp/images/pokemon1.jpg?itok=WnaWpA5f"  height ="200" width = "300">
<p>

En cada entrega podemos encontrar una gran diversidad de pokémon, cada uno con su propio tipo o tipos, acceso a gran diversidad de movimientos especiales y estrategias. Para lograr un sistema de combate lo más equilibrado posible que permitiera combinar todos estos aspectos, se le asignó a cada pokémon una seríe de estadísticas relacionadas con ciertos atributos como vida, ataque, defensa, velocidad entre otros, es decir, una forma de reconocer la fuerza de un pokémon y las posibles estrategias con las cuales combatir. 

<p align="center">
<img src="https://www.cpokemon.com/cutenews/data/upimages/platino_arceus_stats.png"  height ="200" width = "300">
<p>

Para un jugador promedio, este tipo de información puede pasar desapercibida y el juego resultará igual de divertido. Sin embargo, la franquicia cuenta con una escena competitiva: [El Campeonato Mundial de Videojuegos Pokémon](https://pokemon.fandom.com/es/wiki/Campeonato_mundial_de_videojuegos_Pok%C3%A9mon).

<p align="center">
<img src="https://static.wikia.nocookie.net/espokemon/images/9/94/Campeonato_mundial_de_videojuegos.png/revision/latest/scale-to-width-down/600?cb=20170228002738
"  height ="300" width = "300">
<p>

Donde los jugadores entablan combates con sus pokémon más fuertes, resultados de un fuerte análisis de los datos de cada pokémon, así como un profundo entendimiento de las mecánicas del juego y el sistema de entrenamiento. He aquí la importancia de las estadísticas de cada pokémon, con ellas podemos encontrar a los más fuertes o a los más debiles, a los más rápidos o a los más grandes. El mundo pokémon tiene grandes sopresas que estamos a punto de descubrir.



# Sobre el dataset

Como primer paso, conozcamos mejor la información que contiene nuestro dataset. El autor [Mario Tormo Romero](https://www.kaggle.com/mariotormo) es un científico de datos alemán que tomó como inspiración la afición de la franquicia de su hijo al buscar un proyecto con el que realizar análisis exploratorio de datos. Así conforme se encontró con algunos datasets se dio cuenta de que la mayoría estaban desactualizados o tenian incompletos algunos datos relacionados a las estadísticas individuales de cada pokémon. Entonces decidió expandir su proyecto para incluir un poco de minería de datos web.

<p align="center">
<img src="https://storage.googleapis.com/kaggle-avatars/images/4285867-kg.jpg"  height ="200" width = "200">
<p>


Utilizando Python Web Crawler recolectó y organizó datos de la [base de datos pokémon](https://pokemondb.net/) y [serebii](https://www.serebii.net/) de los más de 800 pokémon que existían en el momento que publicó el dataset por primera vez, recientemente actualizó por última vez el dataset el 16 de Abril del 2021 con datos de la última expansión del juego más reciente de la franquicia: [Pokémon Sword & Shield](https://swordshield.pokemon.com/es-mx/). Mario menciona que con este dataset espera podamos practicar y mejorar nuestras habilidades de análisis de datos.


<p align="center">
<img src="https://images-na.ssl-images-amazon.com/images/I/61rALXW-dSL._SL1280_.jpg"  height ="200" width = "250">
<p>




# ¿Qué información podemos encontrar en un dataset sobre pokémon?

El dataset contiene información de 898 pokémon (1044 incluyendo formas regionales) en 51 campos y está organizada en 5 grupos:

<p align="center">
<img src="https://powergamingnetwork.com/wp-content/uploads/2018/05/013.png"  height ="200" width = "250">
<p>

- Datos de la pokédex: Información que se despliega generalmente en la entrada de la pokédex de los juegos.
      - Número de la pokédex: Número de entrada en la pokédex nacional.
      - Nombre: El nombre del pokémon en inglés (global).
      - Nombre en alemán: El nombre del pokémon en alemán.
      - Nombre en Japónes: El nombre del pokémon en japónes.
      - Generación: El número de la generación (1-8) donde se introdujo al pokémon por primera vez.
      - Estatus: Denota si el pokémon es legendario, sub-legendario, mítico o normal.
      - Especie: Indica la categoría del pokémon.
      - Número de tipos: Número de tipos que el pokémon tiene (1-2).
      - Tipo 1: El tipo primario del pokémon.
      - Tipo 2: El tipo secundario del pokémon (sí cuenta con uno).
      - Altura: Altura del pokémon en metros.
      - Peso: Peso del pokémon en kilogramos.
      - Número de habilidades: Número de habilidades a las que el pokémon puede acceder. 
      - Habilidad 1: Nombre de la primer habilidad que el pokémon puede adquirir.
      - Habilidad 2: Nombre de la segunda habilidad que el pokémon puede adquirir (Sí cuenta con una).
      - Habilidad oculta: Nombre de la habilidad oculta que el pokémon puede adquirir (Sí cuenta con una).


<p align="center">
<img src="http://pkmn.net/games/emerald/lightball2.PNG"  height ="200" width = "250">
<p>


- Estadísticas base: Indican la fortaleza de cada pokémon y varian de acuerdo al nivel o forma evolutiva, por esto se toman en cuenta al nivel 1 en cualquier forma.
      - Puntos totales: Suma ponderada de los puntos base.
      - Puntos de vida: Vida base del pokémon.
      - Ataque: Ataque base del pokémon.
      - Defensa: Defensa base del pokémon.
      - Ataque especial: Ataque especial base del pokémon.
      - Defensa especial: Defensa especial base del pokémon.
      - Velocidad: Velocidad base del pokémon.

<p align="center">
<img src="https://i1.wp.com/www.sopitas.com/wp-content/uploads/2016/10/bulbasaur-captura.gif"  height ="200" width = "250">
<p>

- Entrenamiento: Datos relacionados con la posibilidad de capturar, mejorar las estadísticas y evolucionar de un pokémon. 
       - Ratio de captura: Parámetro utilizado para calcular la probabilidad de éxito al capturar un pokémon.
       - Amistad: Amistad base del pokémon.
       - Experiencia base: Experiencia base del pokémon en estaso salvaje al ser capturado.
       - Ratio de crecimiento: Cantidad de experiencia necesaria para subir de nivel.

<p align="center">
<img src="https://cdn2.bulbagarden.net/upload/thumb/0/0a/ExpGraphLv100.png/800px-ExpGraphLv100.png"  height ="200" width = "250">
<p>


- Crianza: Datos relacionados con el procedimiento de crianza de pokémon.
       - Número de grupos huevo: Con cuantos grupos huevos puede criar el pokémon.
       - Grupos huevo: Nombre de los grupos huevo compatibles.
       - Porcentaje de género masculino: Porcentaje de pokémon criados que resultan ser de género masculino durante el proceso de crianza (Valor en blanco si el pokémon no posee género).
       - Ciclos huevo: Número de ciclos (255-257 pasos) requeridos para eclocionar un huevo pokémon.

<p align="center">
<img src="https://64.media.tumblr.com/tumblr_m1m6ws6hsz1r2mho4o1_500.gif"  height ="200" width = "250">
<p>

- Debilidad de tipo: Efectividad de los ataques de cierto tipo contra la combinación de tipos del pokémon.
       - En contra de: 18 elementos que denotan la cantidad de daños recibido por ataques de cierto tipo.

<p align="center">
<img src="https://i.pinimg.com/originals/0d/eb/1d/0deb1d9f5c679deba21c869bb6a66e83.png"  height ="200" width = "250">
    


# Limpieza del dataset

Ahora que sabemos lo que podemos esperar del dataset, debemos elegir los campos más adecuados para un análisis y verificar que no tengamos datos nulos o incompletos.

In [1]:
import pandas as pd
#Importamos una copia desde nuestro repositorio en github
df = pd.read_csv("https://raw.githubusercontent.com/JMCastle98/M4_Analisis_de_Datos_con_Python/main/Dataset/pokedex.csv", index_col=0)

In [2]:
df.head(4)

pokedex_number           name  ... against_steel against_fairy
0               1      Bulbasaur  ...           1.0           0.5
1               2        Ivysaur  ...           1.0           0.5
2               3       Venusaur  ...           1.0           0.5
3               3  Mega Venusaur  ...           1.0           0.5

[4 rows x 50 columns]

Observamos que las primeras 3 entradas de nuestro ~~pokédex~~ dataset corresponden a la línea evolutiva de Bulbasaur y se muestran los campos antes descritos. Realizar el análisis con 51 campos es una tarea complicada por lo que decidimos retirar algunos como:

- Nombres en alemán y japónes. 
- Campos relacionados a las habilidades del pokémon.
- Experiencia base y amistad base. 
- Campos de crianza.
- Campos de debilidades.

Pensamos que estos datos no nos son del todo útiles para el tipo de análisis que queremos realizar y por ello nos quedamos con los siguientes:


In [3]:
df_aux = df[['pokedex_number',
               'name','generation',
               'status', 'species', 
               'type_number', 'type_1', 'type_2', 
               'height_m', 'weight_kg',
               'total_points','hp','attack',
               'defense','sp_attack','sp_defense',
               'speed','catch_rate','growth_rate']]

Observemos el dataframe resultante:

In [4]:
df_aux

pokedex_number                  name  ...  catch_rate  growth_rate
0                  1             Bulbasaur  ...        45.0  Medium Slow
1                  2               Ivysaur  ...        45.0  Medium Slow
2                  3              Venusaur  ...        45.0  Medium Slow
3                  3         Mega Venusaur  ...        45.0  Medium Slow
4                  4            Charmander  ...        45.0  Medium Slow
...              ...                   ...  ...         ...          ...
1040             896             Glastrier  ...         3.0         Slow
1041             897             Spectrier  ...         3.0         Slow
1042             898               Calyrex  ...         3.0         Slow
1043             898     Calyrex Ice Rider  ...         3.0         Slow
1044             898  Calyrex Shadow Rider  ...         3.0         Slow

[1045 rows x 19 columns]

¡Mucho mejor! Ahora tenemos los campos suficientes para describir las características y fortalezas de un pokémon. Sin embargo... ¿Qué es eso que veo? ¡¿Acaso son NaNs?!

<p align="center">
<img src="https://i.kym-cdn.com/photos/images/newsfeed/001/049/745/e5a.gif"  height ="200" width = "250">
<p>


In [5]:
df_aux.info() 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1045 entries, 0 to 1044
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   pokedex_number  1045 non-null   int64  
 1   name            1045 non-null   object 
 2   generation      1045 non-null   int64  
 3   status          1045 non-null   object 
 4   species         1045 non-null   object 
 5   type_number     1045 non-null   int64  
 6   type_1          1045 non-null   object 
 7   type_2          553 non-null    object 
 8   height_m        1045 non-null   float64
 9   weight_kg       1044 non-null   float64
 10  total_points    1045 non-null   int64  
 11  hp              1045 non-null   int64  
 12  attack          1045 non-null   int64  
 13  defense         1045 non-null   int64  
 14  sp_attack       1045 non-null   int64  
 15  sp_defense      1045 non-null   int64  
 16  speed           1045 non-null   int64  
 17  catch_rate      1027 non-null   f

Efectivamente tenemos unos cuantos valores NaNs en el dataset, haremos lo posible para limpiarlos.

In [6]:
df_aux.isnull().sum()

pokedex_number      0
name                0
generation          0
status              0
species             0
type_number         0
type_1              0
type_2            492
height_m            0
weight_kg           1
total_points        0
hp                  0
attack              0
defense             0
sp_attack           0
sp_defense          0
speed               0
catch_rate         18
growth_rate         1
dtype: int64

Lo primero que salta a la vista son los 492 valores NaN en el tipo secundario de los pokémon, esto debido a que algunos pokemon solo tienen un tipo principal, asi que para rellenar los NaNs estableceremos su segundo tipo como "None".

In [7]:
df_clean = df_aux.copy(deep=True)

In [8]:
df_clean['type_2'] = df_aux['type_2'].fillna('None')

In [9]:
df_clean.isnull().sum()

pokedex_number     0
name               0
generation         0
status             0
species            0
type_number        0
type_1             0
type_2             0
height_m           0
weight_kg          1
total_points       0
hp                 0
attack             0
defense            0
sp_attack          0
sp_defense         0
speed              0
catch_rate        18
growth_rate        1
dtype: int64

Ya que nos encargamos de los NaNs en el tipo secundario, veamos que pasa con el peso:

In [10]:
df_clean[ df['weight_kg'].isnull()]

pokedex_number                 name  ...  catch_rate growth_rate
1033             890  Eternatus Eternamax  ...         NaN        Slow

[1 rows x 19 columns]

Eternatus es un pokemon legendario que en su forma Dinamax crece al menos 5 veces en tamaño, sin embargo, dentro de los metadatos del juego (a.k.a la pokédex) al ser gigantesco su peso viene indicado como **???**. Para no dejar cabos sueltos estableceremos su peso como 5 veces su peso normal, es decir, 4750 kg.   

<p align="center">
<img src="https://www.serebii.net/swordshield/pokemon/890.png"  height ="200" width = "250">
<p>

<p align="center">
<img src="https://www.serebii.net/swordshield/pokemon/890-e.png"  height ="200" width = "250">
<p>

In [11]:
df_clean['weight_kg'] = df_aux['weight_kg'].fillna(4750)

In [12]:
df_clean.isnull().sum()

pokedex_number     0
name               0
generation         0
status             0
species            0
type_number        0
type_1             0
type_2             0
height_m           0
weight_kg          0
total_points       0
hp                 0
attack             0
defense            0
sp_attack          0
sp_defense         0
speed              0
catch_rate        18
growth_rate        1
dtype: int64

El siguiente valor nulo lo encontramos en el ratio de crecimiento:

In [13]:
df_clean[ df['growth_rate'].isnull()]

pokedex_number                          name  ...  catch_rate growth_rate
658             555  Galarian Darmanitan Zen Mode  ...         NaN         NaN

[1 rows x 19 columns]

La forma Zen del Darmanitan de Galar tiene un pequeño error en su ratio de crecimiento, en general todas las formas de un Darmanitan comparten el mismo tipo de crecimiento: "Medium Slow"

<p align="center">
<img src="https://images.gameinfo.io/pokemon/256/555-32.png"  height ="250" width = "250">
<p>

In [14]:
df_clean['growth_rate'] = df_aux['growth_rate'].fillna('Medium Slow')

In [15]:
df_clean.isnull().sum()

pokedex_number     0
name               0
generation         0
status             0
species            0
type_number        0
type_1             0
type_2             0
height_m           0
weight_kg          0
total_points       0
hp                 0
attack             0
defense            0
sp_attack          0
sp_defense         0
speed              0
catch_rate        18
growth_rate        0
dtype: int64

En cuanto a los valores nulos en el el campo "catch_rate" ,para algunos pokémon de la region de Galar, encontramos en [serebii](https://www.serebii.net/) (una de las fuentes originales del autor) que dichos valores existen para los pokémon de la lista, pero el valor no fue añadido o no se agregó correctamente. Arreglaremos esto de forma manual:

In [16]:
df_aux[ df['catch_rate'].isnull()]

pokedex_number  ...  growth_rate
69                52  ...  Medium Fast
100               77  ...  Medium Fast
102               78  ...  Medium Fast
111               83  ...  Medium Fast
144              110  ...  Medium Fast
158              122  ...  Medium Fast
273              222  ...         Fast
320              263  ...  Medium Fast
322              264  ...  Medium Fast
654              554  ...  Medium Slow
657              555  ...  Medium Slow
658              555  ...          NaN
666              562  ...  Medium Fast
723              618  ...  Medium Fast
1017             877  ...  Medium Fast
1029             888  ...         Slow
1031             889  ...         Slow
1033             890  ...         Slow

[18 rows x 19 columns]

In [17]:
#Datos obtenidos de la misma fuente del autor
df_clean.at[69,'catch_rate'] = 255.0
df_clean.at[100,'catch_rate'] = 190.0
df_clean.at[102,'catch_rate'] = 60.0
df_clean.at[111,'catch_rate'] = 45.0
df_clean.at[144,'catch_rate'] = 60.0
df_clean.at[158,'catch_rate'] = 45.0
df_clean.at[273,'catch_rate'] = 60.0
df_clean.at[320,'catch_rate'] = 255.0
df_clean.at[322,'catch_rate'] = 90.0
df_clean.at[654,'catch_rate'] = 120.0
df_clean.at[657,'catch_rate'] = 60.0
df_clean.at[658,'catch_rate'] = 60.0
df_clean.at[666,'catch_rate'] = 190.0
df_clean.at[723,'catch_rate'] = 75.0
df_clean.at[1017,'catch_rate'] = 180.0
df_clean.at[1029,'catch_rate'] = 10.0
df_clean.at[1031,'catch_rate'] = 10.0
df_clean.at[1033,'catch_rate'] = 255.0

Con esto terminamos de limpiar nuestro Dataset y podemos comenzar el análisis.

In [19]:
df_clean.isnull().sum()

pokedex_number    0
name              0
generation        0
status            0
species           0
type_number       0
type_1            0
type_2            0
height_m          0
weight_kg         0
total_points      0
hp                0
attack            0
defense           0
sp_attack         0
sp_defense        0
speed             0
catch_rate        0
growth_rate       0
dtype: int64

In [18]:
df_clean.to_csv('pokédex_clean.csv')

# Postwork 1: Obteniendo estimados de locación y variabilidad
Observemos un poco los campos del dataset para encontrar los datos numericos:

In [21]:
import pandas as pd
df_clean = pd.read_csv("https://raw.githubusercontent.com/JMCastle98/M4_Analisis_de_Datos_con_Python/main/Dataset/pok%C3%A9dex_clean.csv", index_col=0)
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1045 entries, 0 to 1044
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   pokedex_number  1045 non-null   int64  
 1   name            1045 non-null   object 
 2   generation      1045 non-null   int64  
 3   status          1045 non-null   object 
 4   species         1045 non-null   object 
 5   type_number     1045 non-null   int64  
 6   type_1          1045 non-null   object 
 7   type_2          1045 non-null   object 
 8   height_m        1045 non-null   float64
 9   weight_kg       1045 non-null   float64
 10  total_points    1045 non-null   int64  
 11  hp              1045 non-null   int64  
 12  attack          1045 non-null   int64  
 13  defense         1045 non-null   int64  
 14  sp_attack       1045 non-null   int64  
 15  sp_defense      1045 non-null   int64  
 16  speed           1045 non-null   int64  
 17  catch_rate      1045 non-null   f

La mayoría de campos son de tipo entero o flotante, los que aportan más información sobre las características del pokémon son:

- altura
- peso
- puntos totales
- estadísticas
- ratio de captura. 

Para facilitar la tarea de obtener los estimados númericos creamos una función.

In [22]:
def estimados(campo):
    from scipy import stats
    print(f'El promedio de {campo.name} es: '+'{0:.3f}'.format(campo.mean()))
    print(f'La mediana de {campo.name} es: '+'{0:.3f}'.format(campo.median()))
    print(f'La media truncada de {campo.name} es: '+'{0:.3f}'.format(stats.trim_mean(campo, 0.1)))
    print(f'La desviación estándar de {campo.name} es: '+'{0:.3f}'.format(campo.std()))
    print(f'El máximo de {campo.name} es: '+'{0:.3f}'.format(campo.max()))
    print(f'El mínimo de {campo.name} es: '+'{0:.3f}'.format(campo.min()))
    print(f'El rango de {campo.name} es: '+'{0:.3f}'.format(campo.max() - campo.min()))
    print(f'El percentil 25 de {campo.name} es: '+'{0:.3f}'.format(campo.quantile(0.25))+' y el percentil 75 es: '+'{0:.3f}'.format(campo.quantile(0.75)))
    print(f'El rango intercuartílico de {campo.name} es: '+'{0:.3f}'.format(campo.quantile(0.75)-campo.quantile(0.25)))

In [23]:
estimados(df_clean['height_m'])

El promedio de height_m es: 1.374
La mediana de height_m es: 1.000
La media truncada de height_m es: 1.057
La desviación estándar de height_m es: 3.353
El máximo de height_m es: 100.000
El mínimo de height_m es: 0.100
El rango de height_m es: 99.900
El percentil 25 de height_m es: 0.600 y el percentil 75 es: 1.600
El rango intercuartílico de height_m es: 1.000


Muy bien, lo primero que salta a la vista es que nuestro amigo Eternatus es un outlier ya que el posee la altura máxima. En general esto no movió mucho a nuestra media ya que tenemos más de 1000 datos y como podemos comprobar con la mediana y la media truncada; 1 metro de altura es lo normal en los pokémon. Observamos que la desviación estándar es algo grande debido al outlier, pero los percentiles 25 y 75 nos indican que la mayoría de datos están entre 0.6 y 1.6 metros. 

In [24]:
estimados(df_clean['weight_kg'])

El promedio de weight_kg es: 75.694
La mediana de weight_kg es: 29.500
La media truncada de weight_kg es: 41.010
La desviación estándar de weight_kg es: 196.021
El máximo de weight_kg es: 4750.000
El mínimo de weight_kg es: 0.100
El rango de weight_kg es: 4749.900
El percentil 25 de weight_kg es: 9.000 y el percentil 75 es: 70.500
El rango intercuartílico de weight_kg es: 61.500


En el peso encontramos como outlier es el de nuestro amigo Eternatus (no por nada es un pokémon legendario) mientras que el peso máximo otros pokémon es de 999.9 kg Encontramos que un pokémon promedio pesa 41.010 kg y que el más ligero pesa solo 100 gramos.

In [25]:
estimados(df_clean['total_points'])

El promedio de total_points es: 439.353
La mediana de total_points es: 458.000
La media truncada de total_points es: 437.200
La desviación estándar de total_points es: 121.993
El máximo de total_points es: 1125.000
El mínimo de total_points es: 175.000
El rango de total_points es: 950.000
El percentil 25 de total_points es: 330.000 y el percentil 75 es: 515.000
El rango intercuartílico de total_points es: 185.000


Hablando de los puntos totales, nos encontramos con pokémon super fuertes con hasta 1125 puntos, sin embargo con la media truncada encontramos que lo más normal son pokémon con 437.2 puntos y podemos establacer el rango entre 330 y 515 puntos. La desviación estándar es algo aceptable con un valor de 121.993.

In [26]:
estimados(df_clean['hp'])

El promedio de hp es: 70.068
La mediana de hp es: 68.000
La media truncada de hp es: 67.994
La desviación estándar de hp es: 26.671
El máximo de hp es: 255.000
El mínimo de hp es: 1.000
El rango de hp es: 254.000
El percentil 25 de hp es: 50.000 y el percentil 75 es: 82.000
El rango intercuartílico de hp es: 32.000


En cuanto a la vida, observamos que lo normal es que los pokémon tengan 68 puntos, los pokémon más fuertes tendrán como máximo 255 y los más normales estarán entre 50 y 82 puntos de vida.

In [27]:
estimados(df_clean['attack'])

El promedio de attack es: 80.477
La mediana de attack es: 77.000
La media truncada de attack es: 78.869
La desviación estándar de attack es: 32.433
El máximo de attack es: 190.000
El mínimo de attack es: 5.000
El rango de attack es: 185.000
El percentil 25 de attack es: 55.000 y el percentil 75 es: 100.000
El rango intercuartílico de attack es: 45.000


Para un pokémon enfocado a ataques físicos, lo normal sería golpear con alrededor de 78.869 puntos de ataque, los que poseen 190 puntos seguramente son pokémon legendarios. Los percentiles nos indican que el 50% de los pokémon tienen una fuerza entre 100 y 55 puntos, algo entendible considerando que no todos están necesariamente en su forma evolutiva más fuerte.

In [28]:
estimados(df_clean['defense'])

El promedio de defense es: 74.671
La mediana de defense es: 70.000
La media truncada de defense es: 71.943
La desviación estándar de defense es: 31.259
El máximo de defense es: 250.000
El mínimo de defense es: 5.000
El rango de defense es: 245.000
El percentil 25 de defense es: 50.000 y el percentil 75 es: 90.000
El rango intercuartílico de defense es: 40.000


La mejor ofensiva es una buena defensa, algunas batallas pokémon se ganan por resistencia y aguante, aquellos con 250 puntos puede que sean de tipo acero o roca. Para el resto de pokémon una defensa cercana a 72 es lo normal, los percentiles van de 50 a 90 puntos y parece que eso basta para aguantar una salpicadura del poderoso Magikarp. 

<p align="center">
<img src="https://media.tenor.com/images/b4c546e6a9a3fdaff5c0841fa38563c1/tenor.gif"  height ="100" width = "100">
<p>


In [29]:
estimados(df_clean['sp_attack'])

El promedio de sp_attack es: 73.032
La mediana de sp_attack es: 65.000
La media truncada de sp_attack es: 70.382
La desviación estándar de sp_attack es: 32.746
El máximo de sp_attack es: 194.000
El mínimo de sp_attack es: 10.000
El rango de sp_attack es: 184.000
El percentil 25 de sp_attack es: 50.000 y el percentil 75 es: 95.000
El rango intercuartílico de sp_attack es: 45.000


Cuando se habla de ataque especial en el mundo pokémon, hablamos de los poderes característicos relacionados con el tipo del pokémon, como lo es el mítico impactrueno. El ataque especial más alto es de 194 puntos y seguramente tenemos un pokémon psiquíco, al resto de pokémon los podemos encontrar entre 50 y 95 puntos, aquellos que tengan menos seguramente serán de tipo lucha.

<p align="center">
<img src="https://pa1.narvii.com/6797/ccf64a56e65ae3b0a1f4b49948e9b9081b8a0c3e_hq.gif"  height ="200" width = "250">
<p>

In [30]:
estimados(df_clean['sp_defense'])

El promedio de sp_defense es: 72.299
La mediana de sp_defense es: 70.000
La media truncada de sp_defense es: 70.372
La desviación estándar de sp_defense es: 28.099
El máximo de sp_defense es: 250.000
El mínimo de sp_defense es: 20.000
El rango de sp_defense es: 230.000
El percentil 25 de sp_defense es: 50.000 y el percentil 75 es: 90.000
El rango intercuartílico de sp_defense es: 40.000


La defensa especial por lo tanto es muy útil para contrarrestar ataques especiales, los pokémon normalmente tendrán 70 puntos y el rango esta entre 50 y 90 puntos por lo que no todos pueden resistir un enfrentamiento de ese tipo. Los mejores pokémon tanque son aquellos con una buena defensa física y especial.

In [31]:
estimados(df_clean['speed'])

El promedio de speed es: 68.808
La mediana de speed es: 65.000
La media truncada de speed es: 67.461
La desviación estándar de speed es: 30.210
El máximo de speed es: 200.000
El mínimo de speed es: 5.000
El rango de speed es: 195.000
El percentil 25 de speed es: 45.000 y el percentil 75 es: 90.000
El rango intercuartílico de speed es: 45.000


La velocidad puede parecer un parámetro sin sentido en un juego de combates por turnos como lo es pokémon, pero este es de vital importancia, ya que el pokémon más rápido ataca primero. Si nuestro pokémon es muy fuerte y queremos dar el golpe decisivo debemos buscar que también sea veloz, o tal vez busquemos un tanque super lento que resista todos los golpes ... las estrategias con infinitas. En cualquier caso, el pokémon promedio tiene 67.461 puntos de velocidad base y el rango está entre 45 y 90 puntos, superar este rango puede significar atacar primero.

In [32]:
estimados(df_clean['catch_rate'])

El promedio de catch_rate es: 93.075
La mediana de catch_rate es: 60.000
La media truncada de catch_rate es: 84.523
La desviación estándar de catch_rate es: 76.045
El máximo de catch_rate es: 255.000
El mínimo de catch_rate es: 3.000
El rango de catch_rate es: 252.000
El percentil 25 de catch_rate es: 45.000 y el percentil 75 es: 127.000
El rango intercuartílico de catch_rate es: 82.000


Finalmente observamos el ratio de captura, el valor máximo es de 255 puntos y generalmente corresponde a pokémon comúnes o sencillos de capturar. Mientras que los pokémon con 3 puntos son los más difíciles y suelen ser legendarios que darán mucha pelea antes de entrar en la pokébola. 

Este parámetro en realidad es un término más de una fórmula secreta de los videojuegos de pokémon que ha ido cambiando con el tiempo, los fans han encontrado que junto a otros parámetros como vida restante, estado alterado y nivel puede obtener la probabilidad real. 

<p align="center">
<img src="https://img.buzzfeed.com/buzzfeed-static/static/2015-09/10/11/enhanced/webdr01/anigif_enhanced-15027-1441899826-4.gif"  height ="200" width = "250">
<p>

Lo bueno es que los pokémon tienen un ratio de captura de 84.523 puntos si dejamos de lado a los más fuertes y débiles, la mayoría tiene un ratio mayor a 45 y menor a 127 puntos. 

Como hemos podido apreciar, las diferentes estadísticas de cada pokémon los vuelven únicos, en un sentido amplio de la palabra. Como entrenadores nuestro objetivo es encontrar al mejor pokémon para cada batalla analizando sus fortalezas y debilidades, los verdaderos maestros pokémon son aquellos que pueden luchar con cualquiera, por ejemplo, este pequeño Pachirisu que hizó historia en 2014 venciendo a muchos pokémon legendarios y llevando a su equipo a la victoria.

<p align="center">
<img src="https://as.com/meristation/imagenes/2018/08/24/reportajes/1535078626_003205_1535079545_sumario_normal.jpg"  height ="200" width = "300">
<p>
